kevm-pyk/src/kevm_pyk/kproj/evm-semantics/evm-types.md

```diff
@@ -323,7 +323,8 @@ A cons-list is used for the EVM wordstack.
     syntax Bytes ::= "#write" "(" Bytes "," Int "," Int ")" [function]
                    | Bytes "[" Int ":=" Bytes "]" [function, total, klabel(mapWriteRange)]
  // --------------------------------------------------------------------------------------
-    rule #write(WM, IDX, VAL) => padRightBytes(WM, IDX +Int 1, 0) [ IDX <- VAL ]
+    syntax Bytes ::= #buf ( Int , Int ) [function, total, smtlib(buf)]
+    rule #write(WM, IDX, VAL) => WM [ IDX := #buf(1, VAL) ]
```

Comment:
I don't think that we should move the definition of `#buf` to here, but instead just include `BUF` module into `EVM-TYPES` module.

But also not sure that is really the best way. How about we just add `concrete` to the `rule #write(...) => padRightBytes(...)`, so it doesn't simplify out?

```
    rule #write(WM, IDX, VAL) => padRightBytes(WM, IDX +Int 1, 0) [ IDX <- VAL ] [concrete]
```

Then, if it's concrete it will simplify to a fully concrete byte array. But if it's symbolic, it will keep the `#write(...)` symbol.

Reply:
We can add the rule:

```
rule #write(WM, IDX, VAL) => WM [ IDX := #buf(1, VAL) ] [symbolic(WM, IDX, VAL), simplification]
```

to the `BUF` module where `#buf` is defined.

Reply:
Moving the definition here is not clean, yes, and I'd like to avoid if I can. We noticed that `#write` is only used for this one `MSTORE8` opcode and nowhere else so we were considering removing it and just inlining the logic here. My understanding is that the idea to use `#buf` here was based which lemmas we already have.

Reply:
Maybe it would be cleaner, as you suggest, to leave `#write()` unevaluated here, but I'm just worried that would require adding a whole set of lemmas to work with `#write()`. I am trying it out.

Reply:
Fixed.